UI/multitrack-video-output.cpp

```diff
@@ -290,22 +291,36 @@ create_video_encoder(DStr &name_buffer, size_t encoder_index,
 			QTStr("FailedToStartStream.FailedToCreateVideoEncoder")
 				.arg(name_buffer->array, encoder_type));
 	}
-	obs_encoder_set_video(video_encoder, obs_get_video());
 
-	obs_video_info ovi;
-	if (!obs_get_video_info(&ovi)) {
+	obs_video_info ovi_storage;
+	if (!obs_get_video_info(&ovi_storage)) {
 		blog(LOG_WARNING,
 		     "Failed to get obs_video_info while creating encoder %zu",
 		     encoder_index);
 		throw MultitrackVideoError::warning(
 			QTStr("FailedToStartStream.FailedToGetOBSVideoInfo")
 				.arg(name_buffer->array, encoder_type));
 	}
+	obs_video_info *ovi = &ovi_storage;
+
+	const char *view_name = nullptr;
```

Comment:
This is probably safer as std::string.

Reply:
safer in what way? `encoder_config` can't be modified in here (unless someone adds a const cast?), so the pointer (once assigned) stays valid for this scope at least (and it's not used outside of this scope)

Reply:
@tt2468 Could you clarify the concern here?

Reply:
From a cursory glance there are multiple reasons:

1. This is C++ code. By default one should use C++ types and avoid low-level C types at all costs.
2. The `view_name` is derived from `encoder_config.view`, which is a `std::string` already. So the code taps into the internal data of that string instead of using a reference to the original `std::string`, particularly as it is only used for `std::map::find` anyway.

It's precisely the kind of code style that the C++ committee and the language designers tell developers to avoid. Use C++ types and avoid pointers and C types at all costs.

Reply:
@PatTheMav: can you help me translate what you're suggesting? https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f16-for-in-parameters-pass-cheaply-copied-types-by-value-and-others-by-reference-to-const even lists pointers as first option when passing an optional argument into a function (c++ doesn't really support rust's `Option<&str>` or similar unfortunately)

so another option would be to do special handling for empty strings, instead of requiring special handling on a different layer (c pointers, and generally everything c++ is kind of terrible for this unfortunately, since most of these won't noisily fail when you use them wrong, and the compiler also doesn't help you due to lack of borrow checking); if that's the preference here I can make that change

or are you asking for this to be translated into `std::string *view_name = nullptr;` instead?

Reply:
The root cause of this is that the `view` name is optional in `VideoEncoderConfiguration`, but I wonder if this has to be the case:

If OBS is supposed to support multiple views _and_ views carry a name, then the main view should also hold a name and not be identified by "the lack of any name", which is brittle.

So instead the requirement could be: OBS can have multiple views, _every_ view needs to have a unique name, _including_ the main view. This would also allow everyone working with views to clearly identify the views and gives OBS itself the ability to "reserve" view names (and there are already ternary expressions trying to use the fallback name "obs_base", so there clearly _is_ a requirement for the main/base view to have a name).

That also has the potential to solve the following issues:

* `adjust_video_encoder_scaling` doesn't have to accept a mish-mash of actual values _or_ `nullptr`s and use ternary expressions to get some video dimensions or view name from something else.
* In the same vein `adjust_video_encoder_scaling`'s argument list wouldn't need to mix mandatory and logically optional arguments instead of following the good practice of putting mandatory arguments first and optional ones last

So there should be a single list of views to which the "main view" is added and additional views can be added by the frontend API. Then the code could work with the expectation that:

* The list contains at least a single view (not having any views would be an invalid application state)
* Every view (including the main view) carries a name and the same metadata and video info

---

This is all complex stuff and I appreciate the work put into this, but that's even more reason to make it as straightforward as possible and avoid a pile of little workarounds here, there, and everywhere to squeeze in this functionality.